### PR TITLE
Add G_IN_SET, patch our internal users via spatch

### PR DESCRIPTION
**Pull request by cgwalters**
I originally tried to get this into GLib:
https://bugzilla.gnome.org/show_bug.cgi?id=783751

But that looks like it's going to fail due to MSVC. Let's add it here at least
so I can start using it tomorrow and not wait for the MSVC team to catch up.

I renamed `glnx-alloca.h` to `glnx-macros.h` as a more natural collective
home for things from systemd's `macro.h`.

Finally, I used a Coccinelle spatch similar to the one referenced
in the above BZ to patch our uses.

### CODE REVIEW

**Comment by gnomesysadmins** [comment]
Thank you for contributing to libglnx!

libglnx uses Bugzilla for code review.

If you have never contributed to GNOME before make sure you have read the
getting started documentation:
http://www.gnome.org/get-involved

Otherwise please visit
https://wiki.gnome.org/Newcomers
and follow the instructions there to upload your change to Bugzilla.

File: glnx-macros.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+#define _G_INSET_CASE_F_19(CASE, X, ...) CASE(X) _G_INSET_CASE_F_18(CASE, __VA_ARGS__)
+#define _G_INSET_CASE_F_20(CASE, X, ...) CASE(X) _G_INSET_CASE_F_19(CASE, __VA_ARGS__)
+
+#define _G_INSET_GET_CASE_F(_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,_15,_16,_17,_18,_19,_20,NAME,...) NAME
+#define _G_INSET_FOR_EACH_MAKE_CASE(...) \
+  _G_INSET_GET_CASE_F(__VA_ARGS__,_G_INSET_CASE_F_20,_G_INSET_CASE_F_19,_G_INSET_CASE_F_18,_G_INSET_CASE_F_17,_G_INSET_CASE_F_16,_G_INSET_CASE_F_15,_G_INSET_CASE_F_14,_G_INSET_CASE_F_13,_G_INSET_CASE_F_12,_G_INSET_CASE_F_11, \
+                               _G_INSET_CASE_F_10,_G_INSET_CASE_F_9,_G_INSET_CASE_F_8,_G_INSET_CASE_F_7,_G_INSET_CASE_F_6,_G_INSET_CASE_F_5,_G_INSET_CASE_F_4,_G_INSET_CASE_F_3,_G_INSET_CASE_F_2,_G_INSET_CASE_F_1) \
+                   (_G_INSET_CASE_F,__VA_ARGS__)
+
+/* Note: claiming the name here even though it isn't upstream yet
```
Comment (jlebon) [inline glnx-macros.h:81]: We should make that whole section dependent on whether `G_IN_SET` is already defined, no? So it doesn't break once it makes it into glib2.

File: glnx-macros.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+#define _G_INSET_CASE_F_17(CASE, X, ...) CASE(X) _G_INSET_CASE_F_16(CASE, __VA_ARGS__)
+#define _G_INSET_CASE_F_18(CASE, X, ...) CASE(X) _G_INSET_CASE_F_17(CASE, __VA_ARGS__)
+#define _G_INSET_CASE_F_19(CASE, X, ...) CASE(X) _G_INSET_CASE_F_18(CASE, __VA_ARGS__)
+#define _G_INSET_CASE_F_20(CASE, X, ...) CASE(X) _G_INSET_CASE_F_19(CASE, __VA_ARGS__)
+
+#define _G_INSET_GET_CASE_F(_1,_2,_3,_4,_5,_6,_7,_8,_9,_10,_11,_12,_13,_14,_15,_16,_17,_18,_19,_20,NAME,...) NAME
+#define _G_INSET_FOR_EACH_MAKE_CASE(...) \
+  _G_INSET_GET_CASE_F(__VA_ARGS__,_G_INSET_CASE_F_20,_G_INSET_CASE_F_19,_G_INSET_CASE_F_18,_G_INSET_CASE_F_17,_G_INSET_CASE_F_16,_G_INSET_CASE_F_15,_G_INSET_CASE_F_14,_G_INSET_CASE_F_13,_G_INSET_CASE_F_12,_G_INSET_CASE_F_11, \
+                               _G_INSET_CASE_F_10,_G_INSET_CASE_F_9,_G_INSET_CASE_F_8,_G_INSET_CASE_F_7,_G_INSET_CASE_F_6,_G_INSET_CASE_F_5,_G_INSET_CASE_F_4,_G_INSET_CASE_F_3,_G_INSET_CASE_F_2,_G_INSET_CASE_F_1) \
+                   (_G_INSET_CASE_F,__VA_ARGS__)
```
Comment (jlebon) [inline glnx-macros.h:79]: This is pretty ingenious!

File: glnx-macros.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ * are members of a small set.  Normally one has to choose to either use `if (x == val || x == otherval ...)`
+ * or a `switch` statement.  This macro is useful to reduce duplication in the first case,
+ * where one can write simply `if (G_IN_SET (x, val, otherval))`, and avoid the verbosity
+ * that the `switch` statement requires.
+ */
+#define G_IN_SET(x, ...)                          \
+        ({                                      \
+                gboolean _g_inset_found = FALSE;            \
+                /* If the build breaks in the line below, you need to extend the case macros */ \
+                static G_GNUC_UNUSED char _static_assert__macros_need_to_be_extended[20 - sizeof((int[]){__VA_ARGS__})/sizeof(int)]; \
```
Comment (jlebon) [inline glnx-macros.h:99]: This should actually be 21 though, right? 20 is the last case we can successfully handle.

/me runs into https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html

Huh, apparently `char var[0];` is valid, so nevermind!

File: glnx-macros.h (hunk 2 above)
Comment (cgwalters) [reply glnx-macros.h:79]: To be clear I didn't invent this :smile: - see the second commit linked from https://bugzilla.gnome.org/show_bug.cgi?id=783751#c7 which links to https://stackoverflow.com/questions/11761703/overloading-macro-on-number-of-arguments

Comment (jlebon) [reply glnx-macros.h:79]: Haha, gotta love the answer starting with "Simple as:"

**Comment by cgwalters** [comment]
Pushed to master with the `#ifndef G_IN_SET` guard, thanks!

**Pull request closed without merging**